juju/js-libjuju
# WD-693 - Refactor the project

## Description

The Current behavior of the `jujulib` for facade classes is to return back the class of the request facade if the **exact** version is supported by the API, otherwise return `undefined`.
This is causing issues now, where the latest version of facades aren't working on some Juju controllers which result in having the dashboard failing everywhere, in order to fix this issue we would like to fallback to the **latest** supported version by the controller.

### Advantages of this approach

The dashboard will work in all the cases.

### Limitations

If the latest version of the juju dashboard is using a specific functionality that is only available on the latest version of a facade, `jujulib` may return an older version of the facade which doesn't support the functionality that we use.

Which is better that not having the dashboard to work at all!

## Done

There will be part 2 PR that will specifically address the issue mentioned above. This PR is a refactor of the project to make the second PR feasible.

- [x] Better support for Typescript: Mainly Typescript file in the root of `/api`
- [x] Change the structure of the facades: e.g. `/api/facades/ClientV5` -> `/api/facades/client/ClientV5.ts`
- [x] Change the structure of the generator codebase: generator related code was in separate folders in the root of the project, I find the new structure simpler to user and read
- [x] Remove the docs from the project: Each time we generate the facades the docs files get updated, which cause a lot of Git changes to track, instead we can make the docs generation a part of the Github action that deploys the `github.io` pages
- [x] Add a script `reganarate-all.js`: that generates all the facades based on historical `schema.json` from Github
- [x] Remove the date of generating the facades from the header

## TODO

- [ ] File a PR to build the docs with the deployment the Github page
- [ ] File a PR to get the most recent version of the facades that is supported by the Juju controller


## QA

- `npm run build`
- `yarn link`
- Update the imports in the juju-dashboard to work with the new structure
- Make sure that the dashboard is working as expected
